### ISSUE TITLE
[ask | need an advice] Debug vs Production

### ISSUE THREAD

**Issue by RZulfikri** [oney/react-native-webrtc#537]
Hello, i'm working with react-native project using WebRTC.
after few testing I found weird issue.

why webrtc connection immidietly connected when app in debug mode ?
but it take long/more time without debug mode ?

note : 
debug mode mean "Start Remote JS Debugging"

#### Platform information

* **React Native version**: 0.55.3
* **Plugin version**: master
* **OS**: Android / iOS

**Comment by saghul** [comment]
I have no idea. Also, what do you mean by "immediately connected"? At the ICE level? Note that when using the debug mode the code is running on your browser's V8 VM, as opposed to the builtin JSC, so you might be hitting some sweet spot. The WebRTC libraries we provide are compiled in Release mode already, so running your app in debug mode should not have any significant impact when it comes to WebRTC performance.

**Comment by RZulfikri** [comment]
@saghul  "immediately connected" mean it only took 1 - 3 second to fully connected (audio and video connected).

When I "start js remote debugging" in caller and calle app using different devices / network it only took 1 - 3 seconds to conected.
But when I release to production Android x Android took around 20 seconds, Android x Ios 20 - 90 seconds, Ios x Ios took around 40 - 120 seconds.

Any Idea how to optimize this ? Oh ya I'm using peer to peer schema

**Comment by 8BallBomBom** [comment]
Bit unusual, not experiencing this myself.
In production/debug were seeing connection times of around 0.5 seconds using stun and turn servers.
How are you handling the ice candidates?

**Comment by RZulfikri** [comment]
@8BallBomBom is it what you mean ?

```
  if (peerConnection) {
    pc = peerConnection
  } else {
    pc = createPeerConnection();
  }
  if (msg.action === 'local_description') {
    const sdp = msg.data
    if (sdp.type === "offer")
      pc.setRemoteDescription(new RTCSessionDescription(sdp), function () {
        if (pc.remoteDescription.type === "offer")
          pc.createAnswer(function (desc) {
            pc.setLocalDescription(desc, function () {
            // send local_description
            }, error => console.log('Error at set local desc', error));
          }, error => console.log('Error at create answer', error));
      }, error => console.log('Error at set remote desc', error));
  } else {
    const candidate = msg.data
    pc.addIceCandidate(new RTCIceCandidate(candidate))
  }
```

Oh ya, in my case I have 2 app. one always be caller and one always be calle.
I follow this example https://github.com/oney/RCTWebRTCDemo, but I split to caller and calle. (one can createPerrConnection, the other only can answer peerConnection).

**Comment by 8BallBomBom** [comment]
On the Calle side of things, how are you handling candidates?
Trickle Ice typically sends out candidates straight after getting a peerconnection and offer going.
Which then means you'd be sending them candidates from the Caller to the Callee before the Callee has even managed to get started setting up a peerconnection with an answer to the offer.
So generally that then means you'd need to save the candidates being received on the Callee ready to inject once the offer from the Caller has been received.
Pretty much how we are handling it without issues.
On top of the fact you might want to consider stun and turn servers if your connecting over different networks.

**Comment by RZulfikri** [comment]
so you mean, both (caller and calle) must initialize / create peerConnection first before start calling or sending candidates ?

**Comment by RZulfikri** [comment]
So when I apply / inject candidate to peerconnection it need time to apply before I apply or inject new candidate ?

do you have any example project for my reference ?

**Comment by RZulfikri** [comment]
@8BallBomBom is it what you mean ? https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/addIceCandidate do react native webrtc have same behaviour with web based webrtc. ?

**Comment by 8BallBomBom** [comment]
Sadly i can't link you to anything as the projects i'm working on aren't on public repos.
The way we handle it on the Callee side is after createAnswer has been called we then run a function which cycles through the stored candidates that were received before the Callee got the offer.

Pretty much this.
```
	if ( remoteCandidates.length < 1 ) { return; };

	for ( let candidate of remoteCandidates ) {
		let iceCandidate = new RTCIceCandidate( candidate );

		peerConnection.addIceCandidate( iceCandidate );
	};

	remoteCandidates = [];
```
That would add the candidates and then clear them out.
We also check when candidates are received and if a peerconnection exists we just straight up call addIceCandidate, otherwise they get stored in remoteCandidates ready to be processed once the offer has arrived.

**Comment by RZulfikri** [comment]
thank you @8BallBomBom  I'll try.

**Comment by RZulfikri** [comment]
@8BallBomBom  Thank you so much, its working. It's better now.

**Comment by 8BallBomBom** [comment]
Glad to hear it :+1:

**Comment by ForevereBest** [comment]
Unable to trigger addIceCandidate,

this is my code 


pc.addIceCandidate(new RTCIceCandidate(message.candidate),()=>{
                console.log('addIceCandidate success');
            },logError())

**Comment by 8BallBomBom** [comment]
@ForevereBest Might be a better idea opening a new issue with more details and an error report.

**Comment by tgreco** [comment]
@8BallBomBom are you doing your addIceCandidate after setLocalDescription(answer) ?
If so why does it need to occur after?
If you have the RTCPeerConnection already instantiated is there any issue with adding ice candidates before the answer is created?

> Sadly i can't link you to anything as the projects i'm working on are private.
> The way we handle it on the Callee side is after createAnswer has been called we then run a function which cycles through stored candidates that were received before the Callee got the offer.
> 
> ```
> if ( remoteCandidates.length < 1 ) { return; };
> 
> for ( let candidate of remoteCandidates ) {
> 	let iceCandidate = new RTCIceCandidate( candidate );
> 
> 	peerConnection.addIceCandidate( iceCandidate );
> };
> 
> remoteCandidates = [];
> ```
> 
> That would add the candidates and then clear them out.
> We also check when candidates are received and if the peerconnection exists then just straight up call addIceCandidate.
> Otherwise they get stored in remoteCandidates ready to be processed once the offer has arrived.

**Comment by 8BallBomBom** [comment]
@tgreco Yes we have the candidates stacking up and then being added after the answer description has been set.
I haven't tried adding them before but i'm not entirely sure that you can actually do that nicely without issues.
As far as i'm aware you'd have to munge the SDP and add in the candidates.
The whole reason we process candidates after the answer was due to trickle ice standards.

**Comment by tgreco** [comment]
@8BallBomBom I appreciate your quick response. 
I am going to try putting this into practice and see what happens.

I was not queueing my ice candidates before the answer was accepted but rather adding them before so.
This made for unpredictable behavior where I would sometimes get audio but most of the times not.

**Comment by tgreco** [comment]
@8BallBomBom so I have the queue setup and I don't process the ICE candidate queue until setLocalDescription has been called.

I was able to get this working perfectly if both devices are on the same network.
However if either device is not on the same network the audio does not come through (I am only doing audio no video).

I thought maybe I needed a TURN server so I used Twilio's but I am still getting no audio output unless on the same network.

I was wondering if you ever had such an issue or have a suggestion to troubleshoot this?
